backend/src/main/java/ai/verta/modeldb/lineage/LineageServiceImpl.java

```diff
@@ -299,33 +325,91 @@ private boolean filterLineageEntry(
+        } catch (InvalidProtocolBufferException e) {
+          LOGGER.error("Unexpected error {}", e.getMessage());
+          Status status =
+              Status.newBuilder()
+                  .setCode(com.google.rpc.Code.INTERNAL_VALUE)
+                  .setMessage(e.getMessage())
+                  .addDetails(Any.pack(LineageEntryBatchResponse.getDefaultInstance()))
+                  .build();
+          throw StatusProto.toStatusRuntimeException(status);
+        }
+      }
+    }
+    return result;
+  }
+
+  private void validate(Session session, Set<String> experimentRunIds)
+      throws ModelDBException, InvalidProtocolBufferException {
+    LinkedList<String> experimentRunIdsList = new LinkedList<>(experimentRunIds);
+    List<ExperimentRunEntity> experimentRunEntities =
+        experimentRunDAO.getExperimentRunEntitiesBatch(session, experimentRunIdsList);
+    if (experimentRunEntities.size() != experimentRunIdsList.size()) {
+      throw new ModelDBException("Can't find all experiment runs");
+    }
+    for (ExperimentRunEntity experimentRunEntity : experimentRunEntities) {
+      if (experimentRunEntity == null) {
+        LOGGER.warn(EXPERIMENT_RUN_NOT_FOUND);
+        throw new ModelDBException(EXPERIMENT_RUN_NOT_FOUND, Code.NOT_FOUND);
+      }
+      roleService.validateEntityUserWithUserInfo(
+          ModelDBServiceResourceTypes.PROJECT,
+          experimentRunEntity.getProject_id(),
+          ModelDBServiceActions.READ);
     }
   }
 
   private void validate(
       Session session,
-      Set<String> experimentRuns,
+      Set<String> experimentRunIds,
       Map<Long, RepositoryContainer> repositories,
       LineageEntry lineageEntry)
       throws InvalidProtocolBufferException, ModelDBException, NoSuchAlgorithmException {
+    validate(session, experimentRunIds, repositories, lineageEntry, null);
+  }
+
+  private boolean validate(
+      Session session,
+      Set<String> experimentRunIds,
+      Map<Long, RepositoryContainer> repositories,
+      LineageEntry lineageEntry,
+      Set<String> filteredExperimentRunIds)
+      throws InvalidProtocolBufferException, ModelDBException, NoSuchAlgorithmException {
     switch (lineageEntry.getDescriptionCase()) {
       case EXPERIMENT_RUN:
-        String experimentRun = lineageEntry.getExperimentRun();
-        if (!experimentRuns.contains(experimentRun)) {
-          ExperimentRunEntity experimentRunEntity =
-              experimentDAO.getExperimentRun(session, experimentRun);
-          experimentRuns.add(experimentRun);
-          // Validate if current user has access to the entity or not
-          roleService.validateEntityUserWithUserInfo(
-              ModelDBServiceResourceTypes.PROJECT,
-              experimentRunEntity.getProject_id(),
-              ModelDBServiceActions.READ);
+        experimentRunIds.add(lineageEntry.getExperimentRun());
+        if (filteredExperimentRunIds == null) {
```

Comment:
if list is empty then what is expected? if it is empty and still it should be return true then if looks like below suggestion
```suggestion
        if (filteredExperimentRunIds == null || filteredExperimentRunIds.isEmpty()) {
```

Reply:
I think if empty it should return false